### PR TITLE
[videolibrary] GUI configuration of video artwork whitelist, including one-touch "all" / "limited"

### PR DESCRIPTION
**Pull request by rmrector**
## Description
Move the video library artwork whitelist to the GUI, simplified UX for "all" and "limited" options. Video counterpart to #18173.

## Motivation and Context
A better out of the box behavior for video library artwork.
screenshot003

screenshot002

### CODE REVIEW

**Comment by rmrector** [comment]
jenkins build this please

**Comment by DaveTBlake** [comment]
I am currently working on this for Music, see if we had similar approaches

**Comment by KarellenX** [comment]
if you create a build I will test @rmrector 

Same goes for you @DaveTBlake

**Comment by rmrector** [comment]
> Will it just pick up all image files in the folder?

Yes, if their name is valid for an art type - alphanumeric only, 25 characters or less.

**Comment by rmrector** [comment]
Win64 build: http://mirrors.kodi.tv/test-builds/windows/win64/KodiSetup-20200618-0c963e41-no-whitelist-all-artwork-x64.exe

**Comment by KarellenX** [comment]
Thanks @rmrector

It picks up all artwork, regardless of name (abiding by your file naming rule above). So all these non standard Kodi image files are scanned and cached. This artwork type is cumulative so when checking the Choose Art screen for any movie or tv show it shows **all** the Art Types. This could potentially amount to hundreds of entries in the Choose Art dialogue screen. 

But it seems that if I add back my advancedsettings.xml file, it prevents scanning of spurious image files.

If I scan a movie or tv show with local artwork, then I add more artwork and Refresh the movie, the new artwork is not picked up. Even removing from the library and scanning the title back does not pick up the new artwork. After a restart and then a Refresh, the new images are found.

Also noticed that the Extract Thumbnail setting is enabled by default. Not sure if this was always the case, but could you disable it by default. I had thumbnails created for all my movies and I know this particular art type is problematic in movies as we have discussed previously especially when using Chorus2.

TV Shows display an empty art type...
EmptyArtType


Movies and TV shows display an icon image. Not sure where that is coming from
icon


I'll keep testing as I come up with different scenarios. Thanks.

**Comment by DaveTBlake** [comment]
It is a huge coinsidence (and useful) that after all this time both of us happen to be looking at the "out the box" artistwork/scraping/whitelist improvements @rmrector, I'm sure @ronie will be pleased (and thanks for your patience ronie and not nagging). 

I had hesitations back then about the default (no whitelist) behaviour becoming to pick up all local image files it finds, as @KarellenX observed that can create a lot of unwanted art types and cached art. Music folders certainly have image files that users don't want loaded into Kodi (cluttering both cache and db), not only created by the user but lower res image copies put there by other media players.

So I have a slightly different proposal:

Control the extra art that is picked up (added to db and cached) during scanning and scraping (separate things for music, one thing for video) with a whitelist of art types.
- No whitelist, then  no extra art is picked up from local files but all the different art types in the scraper art are applied, taking the first URL of each aspect as that art.
- Whitelist, only those art types get automatically added. Pick up those art types from local image files with that name, then fill gaps in images from the scraped list of possible remote art. Not all aspects from the list of URLs may get used.

This means that the default (out the box) set-up will magically get whatever art the scraper provides (first of those possible), but not run amuck picking images the user has left with their media folders.  Using local art is a more deliberate act, after all the images don't get there by magic, and users with extra local art are both more likely to want to include/exclude certain things and be able to set up a whitelist. 

The whitelist is only about the automated scanning/scraping of extra art (folder thumb always being applied), art addons such as Artwork Beef can manage art entirely differently - fetching images and adding art types as it needs. That how you see it?

File: xbmc/video/VideoInfoScanner.cpp
Comment (DaveTBlake) [inline xbmc/video/VideoInfoScanner.cpp:1571]: The art type has to be in art types
` find(artTypes.begin(), artTypes.end(), aspect) != artTypes.end()`

Not really reviewing, but happened to notice this typo

**Comment by ronie** [comment]
just an idea, why not turn it into a blacklist?
similar to the excludefromscan / excludefromlisting advancedsettings we have.

**Comment by rmrector** [comment]
> If I scan a movie or tv show with local artwork, then I add more artwork and Refresh the movie, the new artwork is not picked up. Even removing from the library and scanning the title back does not pick up the new artwork. After a restart and then a Refresh, the new images are found.

Ya, it caches the directory list. It looks like the original code uses a cache as well, so presumably it behaved the same.

> Also noticed that the Extract Thumbnail setting is enabled by default. Not sure if this was always the case, but could you disable it by default.

I hadn't changed it, and making that change would belong in another PR. The actual problems need to be fixed, though, I don't really think the existing behavior should be changed for that reason.

> TV Shows display an empty art type...

Fixed in the next push.

> Movies and TV shows display an icon image. Not sure where that is coming from

That is coming from unrelated changes around the 'icon' image code, and shows up in Kodi 19 master without this PR.

@DaveTBlake Some folks don't let those other applications "run amuck" in the media folders in the first place.

> Using local art is a more deliberate act, after all the images don't get there by magic, and users with extra local art are both more likely to want to include/exclude certain things and be able to set up a whitelist.

I come to the opposite conclusion. I have nicely curated artwork so I want Kodi to pick it all up, just like it picks up all media items I point it at by default. I don't want to have to fiddle with configuration, and I shouldn't have to because my collection is nicely curated.

> The whitelist is only about the automated scanning/scraping of extra art (folder thumb always being applied), art addons such as Artwork Beef can manage art entirely differently - fetching images and adding art types as it needs. That how you see it?

The classic art types "thumb" and "fanart" and "poster" and "banner" don't deserve special treatment and at some point nearly all the code that treats them specially should be removed - but for the alternate file names.

I don't know what Artwork Beef has to do with this.

**Comment by DaveTBlake** [comment]
> Some folks don't let those other applications "run amuck" in the media folders in the first place.

**Comment by DaveTBlake** [comment]
> Some folks don't let those other applications "run amuck" in the media folders in the first place.

Sure, some folks... others will have image files there and no expectation that Kodi is going to start grabbing them all, previous Kodi versions have not done so. Nor should they be expected to create a blacklist to stop it happening.

> I have nicely curated artwork so I want Kodi to pick it all up, just like it picks up all media items I point it at by default.

And you are an experienced Kodi user that wrote an art addon. How many new users, and this is a discussion about out-the-box behaviour,  have their media that well curated?

>The classic art types "thumb" and "fanart" and "poster" and "banner" don't deserve special treatment and at some point nearly all the code that treats them specially should be removed - but for the alternate file names.

Agreed, but for the alternate file names. Using alternate names for album thumb is going to be widespread because Picard calls them cover.jpg and lots of users will have left that, and hence not a feature that can just be dropped. Many will call it folder.jpg, but even that breaks the art type  as filename idea. The removal of the treatment of fanart separately is lower hanging friut, but needs db changes too. I don't know about the video side.

**Comment by rmrector** [comment]
Fair enough. There at least must be a way to configure it to add all artwork. That is even something I am capable of putting in the GUI settings.

**Comment by DaveTBlake** [comment]
>There at least must be a way to configure it to add all artwork. 

Yes I was thinking that a "grab all image files and make art types" setting is what is needed. 

I would like to have all the music art related settings on a dialog (have run out of settings levels) rather than making the scrolling on the system>media>music settings screen even longer. I have just started to look at that - move the whitelists out of as.xml onto GUI along with thumb files options, and other image source priority options. Plus a "grab all image files" setting.

Then the choice of what is best default behaviour is just about modifying default settings.

**Comment by rmrector** [comment]
I like the idea, I'll take your lead on the more comprehensive GUI configuration.

**Comment by rmrector** [comment]
I implemented the "None" artwork option as the first step of `CVideoThumbLoader::GetArtTypes`, so it also applies to artwork shown / cached while browsing the file system. Kodi still extracts thumbnails from video files to fill the "thumb" artwork type if that option is enabled.

This reuses a handful of language strings from #18173 - my plan is to rebase once that is merged to consolidate strings and add migration from advanced settings.

**Comment by KarellenX** [comment]
Is it possible to add a setting to allow users to download the original artwork to their local folders?

**Comment by rmrector** [comment]
Sure it's possible, but there are a lot of things to consider - "adding the setting" is the tip of the iceberg. I've been trying to come up with something that can be done for v19, that is the last thing Artwork Beef is really good for.

File: system/settings/settings.xml
Comment (DaveTBlake) [inline system/settings/settings.xml:1003]: Watchout on rebase,  #39120 became #39123

**Comment by DaveTBlake** [comment]
> Is it possible to add a setting to allow users to download the original artwork to their local folders?

> I've been trying to come up with something that can be done for v19, that is the last thing Artwork Beef is really good for.

My thoughts are to make saving the original an optional part of caching, hence only happens when remote art is first used by a skin, so avoiding at level = max the fetching and storing of scraped art results that the user doesn't really want or ever even see. And only one request is made for the image from the remote site.

**Comment by rmrector** [comment]
Rebased on #18173 - added AS.xml setting migration and taking Dave's lead on art type "families" configuration.

**Comment by DaveTBlake** [comment]
@ronie , @KarellenX could you take this final version out for a spin to ensure some gremlin hasn't crept in.

**Comment by ronie** [comment]
i've tested current master + AS whitelist & this PR without AS whitelist and the results are the same.
so, looks good to me. but i'm not that much into the video side of Kodi, so don't take my word for it.

**Comment by KarellenX** [comment]
I ran a few test on this build and came up with the following...

### General

1. There is no Movie Set whitelist `<moviesetextraart>`
2. There is no Season white list `<tvseasonextraart>`

The settings for both the above were not transferred out of the advancedsettings.xml file.

### Test 1- Fresh Install
This is a fresh install, no as.xml but movie set artwork folder added.

Movies- All artwork in folder added, including random artwork Kodi would not use
Movie Sets- All artwork from the artwork Movie Set Artwork folder was added

There is an inconsistency between the **Movie Set** Information Page Choose Art  and the **Movie** Information Page Choose Art.
The **Movie Set** Information Page Choose Art has scanned the artwork from the Movie Set Artwork Folder
The **Movie** Information Page Choose Art seems to have scraped different artwork and ignored the Artwork folder.

moviesetart1


**TV Shows & Episodes** - All artwork scanned


### Test 2- advancedsettings.xml file present
Fresh install, no adjustments except to add the advancedsettings.xml file.
Nominated artwork in the advancedsettings.xml file were transferred to the Custom level.

Scraping movies- all artwork was loaded including multiple fanart/posters.
Scraping tv shows and episodes- All artwork loaded including multiple fanart/posters.


### Test 3- Basic
Fresh setup, no advancedsettings.xml file

Movies- Poster and Fanart only scanned. Also multiple poster and multiple fanart. Is this expected?
Move Set- Poster, fanart and multiple poster/fanart scanned

TV Shows- Banner, poster, fanart and multiple poster/fanart scanned
Seasons- unknown
Episodes- thumb only. episode fanart not scanned.


### Test 4- None
**Movies**

- Actors images scanned/scraped. I guess the user has to disable the setting "Download actors images" in addition to setting None in the Artwork level setting? Even after disabling this setting. Some actor images slipped through, but not all of them.
- thumbnail image extracted. Same as above, need to disable "Extract thumbnail image" in addition? Yes, disabling this setting prevents thumbnail being generated. 

**Movie Sets**
No images scanned

**TV Shows**
No images scanned except an actors image that slips through for some reason.

**Comment by manfeed** [comment]
Will the new options for adding extended artwork work for gif files? I was hoping that in Matrix they would be treated like any other kind of artwork, just naming them animatedposter or animatedfanart... Thanks!

**Comment by DaveTBlake** [comment]
@KarellenX regarding your general point, movies and movie sets share a whitelist as do TVshows and seasons, it says so in the setting descriptions and this is by design. This seems reasonable to me, @rmrector mentioned it here https://github.com/xbmc/xbmc/pull/18072#issuecomment-678602669

Did some more tests to see if I could confirm or clarify other things.

**artlevel = basic** 
I may have misunderstood, but I did think that files called poster.jpg and fanart.jpg would be picked up as poster and fanart etc., but it seems that the movie name prefix is mandatory. Has that always been necessary? Confirm that files like <showname>-poster3.jpg etc. gets picked up too. @rmrector I know that it my type family idea biting you, but it sould be possible to check level and avoid it. Music avoids the issue because all basic art types are handled separately from the whitelist (for now, I have work on that).

**artlevel = none** 
For TVShows, actors thumbs are getting picked up from the scraper results and added to the art table (and some cached in texture db but not all, probably server access). Moreover I also got records in the art table for episode "thumb" and tvshow "thumb" but _without a URL_. The show media folders have local <showname>-thumb.jpg files, I think these are related.

Level = maximum and custom seem to be working as expected, unable to test movie sets.

Not sure if you intend to squash commits, otherwise the commit mesaages could do with editing e.g. "code bunk"

**Comment by KarellenX** [comment]
>This seems reasonable to me, @rmrector mentioned it here #18072 (comment)

Ah, yes. Hidden in plain sight. :)


>I did think that files called poster.jpg and fanart.jpg would be picked up as poster and fanart etc., but it seems that the movie name prefix is mandatory.

Not mandatory. Mr Music Man, you did not enable the setting _Movies are in separate folders that match the movie title_ which is what allows you to use the short name format.
https://kodi.wiki/view/Adding_video_sources#Set_Content

File: xbmc/video/VideoThumbLoader.cpp
Comment (DaveTBlake) [inline xbmc/video/VideoThumbLoader.cpp:309]:
```
 if (CServiceBroker::GetSettingsComponent()->GetSettings()->GetInt(
          CSettings::SETTING_VIDEOLIBRARY_ARTWORK_LEVEL) ==
      CSettings::VIDEOLIBRARY_ARTWORK_LEVEL_CUSTOM)
    StringUtils::TrimRight(compareArtType, "0123456789");
```
Will avoid the pickup of families of art types e.g. fanart1, fanart2 etc. when artlevel is basic

File: xbmc/video/VideoInfoScanner.cpp
Comment (DaveTBlake) [inline xbmc/video/VideoInfoScanner.cpp:1519]: Starting `GetArtwork` with
```
   int artworkLevel = CServiceBroker::GetSettingsComponent()->GetSettings()->GetInt(
      CSettings::SETTING_VIDEOLIBRARY_ARTWORK_LEVEL);
    if (artworkLevel == CSettings::VIDEOLIBRARY_ARTWORK_LEVEL_NONE)
      return;
```
will skip fetching actors thumbs when art level = none and also the other unnecessary processing.
(Can't insert it as a comment on the line I want, this is near as can get)

File: xbmc/video/VideoInfoScanner.cpp
Comment (DaveTBlake) [inline xbmc/video/VideoInfoScanner.cpp:1534]: Which gives this
```
bool addAll = (artworkLevel == CSettings::VIDEOLIBRARY_ARTWORK_LEVEL_ALL);
```

**Review by DaveTBlake — COMMENTED** [review]
@rmrector some code suggestions for fixing those couple of issues (unless you can see better).

Also to stop actors thumbs being added when a refresh is done on info dialog `CGUIDialogVideoInfo::SetMovie` needs the settings check too.

```
-      else if (CServiceBroker::GetSettingsComponent()->GetSettings()->GetBool(CSettings::SETTING_VIDEOLIBRARY_ACTORTHUMBS))
-      { // backward compatibility
-        std::string thumb = CScraperUrl::GetThumbUrl(it->thumbUrl.GetFirstUrlByType());
-        if (!thumb.empty())
-        {
-          item->SetArt("thumb", thumb);
-          CTextureCache::GetInstance().BackgroundCacheImage(thumb);
+      else
+      {
+        const std::shared_ptr<CSettings> settings =
+            CServiceBroker::GetSettingsComponent()->GetSettings();
+        if (settings->GetInt(CSettings::SETTING_VIDEOLIBRARY_ARTWORK_LEVEL) !=
+                CSettings::VIDEOLIBRARY_ARTWORK_LEVEL_NONE &&
+            settings->GetBool(CSettings::SETTING_VIDEOLIBRARY_ACTORTHUMBS))
+        { // backward compatibility
+          std::string thumb = CScraperUrl::GetThumbUrl(it->thumbUrl.GetFirstUrlByType());
+          if (!thumb.empty())
+          {
+            item->SetArt("thumb", thumb);
+            CTextureCache::GetInstance().BackgroundCacheImage(thumb);
+          }
         }
```

But I don't have a clear view on how to stop thumbnails being created from the media files themselves when art level = none, and so ensuring no art gets cached

**Comment by rmrector** [comment]
> Test 1

`set` and `setfanart` are misplaced artwork from the file system, NFO, or scraper - Kodi doesn't put that there.


Should this setting really apply to actor images and extracted thumbs? Extracted thumbs have their own specific setting. How about we leave "None" as is for this PR and later it can be decided how much further beyond the base logic this really needs to go.

> EDIT: Those empty thumbs are comming from CVideoInfoScanner::AddVideo and
> // ensure the art map isn't completely empty by specifying an empty thumb
> I'm sure I should remenber why an empty "thumb" is added, but I can't.

There might be some logic that checked for an empty artmap to decide if it needs to do more processing.

**Comment by DaveTBlake** [comment]
>Should this setting really apply to actor images and extracted thumbs? Extracted thumbs have their own specific setting. How about we leave "None" as is for this PR and later it can be decided how much further beyond the base logic this really needs to go.

Yes I think "none" should mean "none". I also think that the extracted thumbs setting (only other video art setting?) should be in the Artwork section and dependant on art level.
But sure leave it to a separate PR.

>There might be some logic that checked for an empty artmap to decide if it needs to do more processing.

Yes there was in several places, but it was removed when the default icon was moved to the art map. I suspect _"to ensure the backward compatibility code doesn't kick in"_ related to something else which may or maybe still be needed. The thumbs in the art table with empty URLs do eventually get thumbnail images from the media file added by the background loader. A user with Kodi on a limited device wanting to turn art off to minimise memory use would want this to stop too. 

Anyway, happy for us to say "beyond this PR", this already achieves enough.

**Review by DaveTBlake — APPROVED** [review]
Thanks @rmrector having retested with those final tweeks this is ready to go. 
Just the wiki to update!

**Comment by KarellenX** [comment]
Is it expected that having **folder.jpg** in a movie folder will be scanned as a poster?

If both **poster.jpg** and **folder.jpg** are in the movie folder, the poster.jpg is used. If poster.jpg is missing, then folder.jpg is used.

**Comment by rmrector** [comment]
Yes, that is existing behavior that was preserved.

**Comment by KarellenX** [comment]
Ok thanks. 

It does make me scratch my head wondering why we have a whitelist, where we can nominate exactly what artwork we want only to have it overridden with the exact artwork type we are trying to avoid being scanned. Users may have a low quality folder.jpg used or added by other software.

**Pull request merged**
